### PR TITLE
Off heap Changes

### PR DESCRIPTION
**Pull request by rmnattas**
Revert "Add APIs to help with array access":
* APIs implementation moved to OMR
* Depends on https://github.com/eclipse/omr/pull/7296

OffHeap changes:
* Use dataAddr field for array access in Walker.cpp 
* Teach DAA about dataAddr pointer 
* Update IdiomRecognitionUtils.cpp to use new TransformUtil APIs 
* Teach J9RecognizedCallTransformer.cpp about dataAddr pointer 
* Update VectorAPIExpansion.cpp to use new TR::TransformUtil APIs 
* Use dataAddr pointer in _jitProfileStringValue 
* Update dereferenceStructPointerChain to use dataAddr field

### CODE REVIEW

File: runtime/compiler/optimizer/J9TransformUtil.hpp [hunk 1]
```diff
@@ -107,91 +107,7 @@ class OMR_EXTENSIBLE TransformUtil : public OMR::TransformUtilConnector
...
-    */
-   static TR::Node *generateArrayOffsetTrees(
-      TR::Compilation *comp,
-      TR::Node *indexNode,
-      TR::Node *elementSizeNode = NULL,
-      int32_t elementSize = 0,
-      bool useShiftOpCode = false);
-
-   static bool transformDirectLoad(TR::Compilation *comp, TR::Node *node);
+   static bool transformDirectLoad(TR::Compilation *, TR::Node *node);
```
Comment (zl-wang) [inline runtime/compiler/optimizer/J9TransformUtil.hpp:110]: any particular purpose in skipping "comp" argument naming? having it is more readable.

File: runtime/compiler/runtime/JitRuntime.cpp [hunk 2]
```diff
@@ -645,6 +645,14 @@ void J9FASTCALL _jitProfileStringValue(uintptr_t value, int32_t charsOffset, int
       {
       readValues = true;
 
+      uintptr_t startOfData = value;
+#if defined(J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+      if (TR::Compiler->om.isOffHeapAllocationEnabled())
+         {
+         startOfData = *((uintptr_t *) (value + TR::Compiler->om.offsetOfContiguousDataAddrField()));
```
Comment (zl-wang) [inline runtime/compiler/runtime/JitRuntime.cpp:652]: offsetOfContiguousDataAddrField is still within the header size, not at the start of data yet.  if "value" is an off-heap array (admittedly unlikely), you cannot get to the start of data this way without loading dataAddr.  This looks incorrect.

**Review by zl-wang — CHANGES_REQUESTED** [review]

File: runtime/compiler/optimizer/J9TransformUtil.hpp (hunk 1 above)
Comment (rmnattas) [reply runtime/compiler/optimizer/J9TransformUtil.hpp:110]: Apparently this is how it was originally before offheap, the commit this is reverting added the "comp". Will modify the revert to keep the "comp" argument name.

File: runtime/compiler/runtime/JitRuntime.cpp (hunk 2 above)
Comment (rmnattas) [reply runtime/compiler/runtime/JitRuntime.cpp:652]: We're dereferencing the sum of value + offsetOfContiguousDataAddrField to get the address of the start of data, or the dereference is incorrect?

File: runtime/compiler/ilgen/J9ByteCodeIlGenerator.hpp [hunk 3]
```diff
@@ -219,6 +219,7 @@ class TR_J9ByteCodeIlGenerator : public TR_IlGenerator, public TR_J9ByteCodeIter
    void         storeArrayElement(TR::DataType dt, TR::ILOpCodes opCode, bool checks = true);
 
    void         calculateElementAddressInContiguousArray(int32_t, int32_t);
+   void         calculateElementAddressInContiguousArray(int32_t);
```
[the remark this answers is not in the document]
Comment (rmnattas) [reply runtime/compiler/ilgen/J9ByteCodeIlGenerator.hpp:222]: Correct, will fix it

File: runtime/compiler/ilgen/Walker.cpp [hunk 4]
```diff
@@ -2056,6 +2060,38 @@ TR_J9ByteCodeIlGenerator::calculateElementAddressInContiguousArray(int32_t width
...
+   push(firstArrayElementAddress);
+   push(index);
+
+   // stack is now ...,firstArrayElement,index<===
+   genUnary(TR::i2l, isForArrayAccess);
+
+   if (shift)
+      {
+      loadConstant(TR::iconst, shift);
+      // stack is now ...,aryRef,firstArrayElement,index,shift<===
```
[the remark this answers is not in the document]
Comment (rmnattas) [reply runtime/compiler/ilgen/Walker.cpp:2086]: Yes, will fix

File: runtime/compiler/runtime/JitRuntime.cpp (hunk 2 above)
Comment (zl-wang) [reply runtime/compiler/runtime/JitRuntime.cpp:652]: ah ... didn't notice the dereferencing

**Comment by zl-wang** [comment]
Jenkins test sanity.functional plinux,xlinux,zlinux jdk8,jdk11

**Comment by zl-wang** [comment]
pLinux machine resource is lacking ...

**Pull request merged**